File: flutter_twilio_commkit_ios/ios/Classes/Core/TwilioEventSink.swift
import Flutter

/// Thread-safe EventChannel sink wrapper for iOS.
/// Buffers events that arrive before Flutter calls onListen, then flushes them.
final class TwilioEventSink: NSObject, FlutterStreamHandler {

    private var eventSink: FlutterEventSink?
    private var pendingEvents: [[String: Any?]] = []

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        // Flush buffered events
        let toFlush = pendingEvents
        pendingEvents.removeAll()
        toFlush.forEach { events($0) }
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }

    /// Sends an event map to Flutter. Always dispatches on main thread.
    /// If Flutter is not yet listening, buffers the event for later delivery.
    func send(_ event: [String: Any?]) {
        DispatchQueue.main.async { [weak self] in
            guard let self = self else { return }
            if let sink = self.eventSink {
                sink(event)
            } else {
                self.pendingEvents.append(event)
            }
        }
    }

    func sendError(code: String, message: String) {
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?(FlutterError(code: code, message: message, details: nil))
        }
    }
}

File: flutter_twilio_commkit_android/android/src/main/kotlin/com/twiliocommkit/android/core/TwilioEventSink.kt
package com.twiliocommkit.android.core

import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel

/**
 * Thread-safe EventChannel.EventSink wrapper.
 *
 * Key behaviours:
 * - Buffers events that arrive BEFORE Flutter starts listening (onListen)
 * - Always dispatches on the main thread (required by Flutter platform channels)
 * - Flushes buffer immediately when Flutter listener attaches
 */
class TwilioEventSink : EventChannel.StreamHandler {

    private val mainHandler = Handler(Looper.getMainLooper())

    @Volatile
    private var sink: EventChannel.EventSink? = null

    // Buffer events that arrive before Flutter is listening
    private val pendingEvents = mutableListOf<Map<String, Any?>>()

    override fun onListen(arguments: Any?, eventSink: EventChannel.EventSink) {
        sink = eventSink
        // Flush any buffered events
        val events = synchronized(pendingEvents) {
            val copy = pendingEvents.toList()
            pendingEvents.clear()
            copy
        }
        events.forEach { eventSink.success(it) }
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }

    /**
     * Sends [event] to Flutter. Safe to call from any thread.
     * If Flutter is not yet listening, the event is buffered and sent
     * as soon as [onListen] is called.
     */
    fun send(event: Map<String, Any?>) {
        mainHandler.post {
            val s = sink
            if (s != null) {
                s.success(event)
            } else {
                synchronized(pendingEvents) {
                    pendingEvents.add(event)
                }
            }
        }
    }

    /**
     * Sends an error event to Flutter.
     */
    fun sendError(code: String, message: String, details: Any? = null) {
        mainHandler.post {
            sink?.error(code, message, details)
        }
    }
}
